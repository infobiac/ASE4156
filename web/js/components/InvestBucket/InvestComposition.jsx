// @flow
import React from 'react';
import Dialog, {
  DialogContent,
  DialogTitle,
  DialogActions,
} from 'material-ui/Dialog';
import TextField from 'material-ui/TextField';
import Paper from 'material-ui/Paper';
import Autocomplete from 'react-autocomplete';
import { Range } from 'rc-slider';
import Table, { TableBody, TableCell, TableHead, TableRow } from 'material-ui/Table';
import IconButton from 'material-ui/IconButton';
import DeleteIcon from 'material-ui-icons/DeleteForever';
import AddIcon from 'material-ui-icons/Add';
import Button from 'material-ui/Button';
import 'rc-slider/assets/index.css';

type Chunk = {id: string, quantity: number, value: number, name: string};
type ChunkList = Array<Chunk>;
type Suggestion = {id: string, name: string, value: number};
type Props = {
  chunks: ChunkList,
  total: number,
  chunkUpdate: ChunkList => void,
  suggestionFieldChange: string => void,
  suggestions: Array<Suggestion>,
  saveFunc: () => void,
  cancelFunc: () => void,
}
type State = {
  suggestionText: string,
}

class InvestComposition extends React.Component<Props, State> {
  static makeIntervals(chunks: ChunkList): Array<number> {
    return chunks.reduce((all, v) => {
      all.push(all[all.length - 1] + (v.value * v.quantity));
      return all;
    }, [0]);
  }
  static renderTableHeader() {
    return (
      <TableHead>
        <TableRow>
          <TableCell padding="dense">Delete</TableCell>
          <TableCell padding="dense">Name</TableCell>
          <TableCell padding="dense">Quantity</TableCell>
          <TableCell padding="dense">Value</TableCell>
          <TableCell padding="dense">Total Value</TableCell>
        </TableRow>
      </TableHead>
    );
  }
  static renderActions(cancel: () => void, save: () => void) {
    return (
      <DialogActions>
        <Button onClick={cancel} color="primary">
          Cancel
        </Button>
        <Button onClick={save} color="primary">
          Save
        </Button>
      </DialogActions>
    );
  }
  static renderCurrentAvailable(available: number) {
    return (
      <TableRow>
        <TableCell padding="dense" />
        <TableCell padding="dense">Available</TableCell>
        <TableCell padding="dense" />
        <TableCell padding="dense" />
        <TableCell padding="dense" numeric>{available.toFixed(2)}</TableCell>
      </TableRow>
    );
  }
  constructor() {
    super();
    this.state = {
      suggestionText: '',
    };
  }
  selectedStock = () => {
    const addition = this.props.suggestions.filter(x => x.name === this.state.suggestionText);
    if (addition.length === 0) {
      return null;
    }
    return addition[0];
  }
  addStock = () => {
    const { chunks } = this.props;
    const addition = this.props.suggestions.filter(x => x.name === this.state.suggestionText);
    if (addition.length === 0) {
      return;
    }
    chunks.push({
      quantity: this.calculateAvailable() / addition[0].value,
      ...addition[0],
    });
    this.setState(() => ({ suggestionText: '' }), () => this.props.suggestionFieldChange(''));
    this.props.chunkUpdate(chunks);
  }
  intervalUpdate = (intervals: Array<number>) => {
    const prevIntervals = InvestComposition.makeIntervals(this.props.chunks);
    let diff = -1;
    for (let i = 0; i < prevIntervals.length; i += 1) {
      if (prevIntervals[i] !== intervals[i]) {
        diff = i;
        break;
      }
    }
    if (diff > 0 && diff < prevIntervals.length) {
      const { chunks } = this.props;
      const cPrev = chunks[diff - 1];
      cPrev.quantity = (intervals[diff] - intervals[diff - 1]) / chunks[diff - 1].value;
      chunks[diff - 1] = cPrev;
      if (diff < prevIntervals.length - 1) {
        const cNext = chunks[diff];
        cNext.quantity = (intervals[diff + 1] - intervals[diff]) / chunks[diff].value;
        chunks[diff] = cNext;
      }
      this.props.chunkUpdate(chunks);
    }
  }
  calculateAvailable = () =>
    this.props.total
    - this.props.chunks.reduce((sum, c) => sum + (c.value * c.quantity), 0)
  deleteChunk = (id: string) => () => {
    const chunks = [...this.props.chunks];
    let index = -1;
    for (let ii = 0; ii < chunks.length; ii += 1) {
      if (chunks[ii].id === id) {
        index = ii;
      }
    }
    /* istanbul ignore next */
    if (index >= 0) {
      chunks.splice(index, 1);
    }
    this.props.chunkUpdate(chunks);
  }
  renderChunk = (c: Chunk) => (
    <TableRow key={c.id}>
      <TableCell padding="dense">
        <IconButton id={`delete-chunk-${c.id}`} onClick={this.deleteChunk(c.id)}>
          <DeleteIcon />
        </IconButton>
      </TableCell>
      <TableCell padding="dense">
        {c.name}
      </TableCell>
      <TableCell padding="dense" numeric>{c.quantity.toFixed(2)}</TableCell>
      <TableCell padding="dense" numeric>{c.value}</TableCell>
      <TableCell padding="dense" numeric>{(c.quantity * c.value).toFixed(2)}</TableCell>
    </TableRow>
  )
  renderNewStockNameField = () => (
    <Autocomplete
      id="select-stock"
      getItemValue={item => item.name}
      items={this.props.suggestions}
      renderInput={(props: {ref: any}) => {
        const { ref, ...rest } = props;
        return (
          <TextField id="stockname" inputRef={ref} {...rest} />
        );
      }
      }
      renderItem={(item, isHighlighted) =>
        (
          <Paper style={{ background: isHighlighted ? 'lightgray' : 'white' }}>
            {item.name}
          </Paper>
        )
      }
      value={this.state.suggestionText}
      onChange={(e) => {
          const text = e.target.value;
        this.setState(() => ({
            suggestionText: text,
        }), () => this.props.suggestionFieldChange(text));
      }}
      onSelect={(val) => {
        this.setState(() => ({
            suggestionText: val,
        }), () => this.props.suggestionFieldChange(val));
      }}
    />
  )
  renderAddStockRow = (selectedStock: ?Suggestion, available: number) => (
    <TableRow>
      <TableCell padding="dense">
        <IconButton id="add-stock" onClick={this.addStock}>
          <AddIcon />
        </IconButton>
      </TableCell>
      <TableCell padding="dense">
        {this.renderNewStockNameField()}
      </TableCell>
      <TableCell padding="dense">
        {selectedStock ? (available / selectedStock.value).toFixed(2) : null}
      </TableCell>
      <TableCell padding="dense">
        {selectedStock ? selectedStock.value : null}
      </TableCell>
      <TableCell padding="dense">
        {selectedStock ? available.toFixed(2) : null}
      </TableCell>
    </TableRow>
  )
  render() {
    const values = InvestComposition.makeIntervals(this.props.chunks);
    const available = this.calculateAvailable();
    const selectedStock = this.selectedStock();
    const rowsCurrentStock = this.props.chunks.map(this.renderChunk);
    return (
      <Dialog open>
        <DialogTitle>Edit Composition</DialogTitle>
        <DialogContent>
          <Table>
            {InvestComposition.renderTableHeader()}
            <TableBody>
              {
                rowsCurrentStock
              }
              {
                this.renderAddStockRow(selectedStock, available)
              }
              {
                InvestComposition.renderCurrentAvailable(available)
              }
            </TableBody>
          </Table>
          <Range
            id="range"
            pushable={false}
            onChange={this.intervalUpdate}
            count={this.props.chunks.length}
            value={values}
            max={this.props.total}
            step={0.01}
          />
        </DialogContent>
        {InvestComposition.renderActions(this.props.cancelFunc, this.props.saveFunc)}
      </Dialog>
    );
  }
}

export default InvestComposition;
