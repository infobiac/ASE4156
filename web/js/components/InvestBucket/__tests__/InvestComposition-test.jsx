import React from 'react';
import { shallow } from 'enzyme';
import InvestComposition from '../InvestComposition';

describe('InvestComposition', () => {
  it('exists', () => {
    expect(InvestComposition).toBeTruthy();
  });

  it('render', () => {
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={jest.fn()}
      suggestionFieldChange={jest.fn()}
      suggestions={[]}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(comp).toMatchSnapshot();
  });

  it('assert text render', () => {
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={jest.fn()}
      suggestionFieldChange={jest.fn()}
      suggestions={[]}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(comp.find('#select-stock').props().renderItem({ name: 'testname' }, true)).toMatchSnapshot();
    expect(comp.find('#select-stock').props().renderItem({ name: 'testname' }, false)).toMatchSnapshot();
  });

  it('assert text render', () => {
    const chgFn = jest.fn();
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={jest.fn()}
      suggestionFieldChange={chgFn}
      suggestions={['AAA']}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(chgFn.mock.calls.length).toEqual(0);
    comp.find('#select-stock').props().onSelect('AAA');
    expect(chgFn.mock.calls.length).toEqual(1);
    comp.update();
    expect(comp.find('#select-stock').props().value).toEqual('AAA');
    expect(chgFn.mock.calls[0][0]).toEqual('AAA');
  });

  it('assert suggestion render', () => {
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={jest.fn()}
      suggestionFieldChange={jest.fn()}
      suggestions={[]}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(comp.find('#select-stock').props().renderInput({ ref: null })).toMatchSnapshot();
  });

  it('assert item value', () => {
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={jest.fn()}
      suggestionFieldChange={jest.fn()}
      suggestions={[]}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(comp.find('#select-stock').props().getItemValue({ name: 'test' })).toEqual('test');
  });

  it('add stock', () => {
    const suggestionFieldChange = jest.fn();
    const chunkUpdate = jest.fn();
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={chunkUpdate}
      suggestionFieldChange={suggestionFieldChange}
      suggestions={[{ id: 1, name: 'AAA', value: 3 }]}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(suggestionFieldChange.mock.calls.length).toEqual(0);
    comp.find('#select-stock').simulate('change', { target: { value: 'AAA' } });
    expect(suggestionFieldChange.mock.calls.length).toEqual(1);
    expect(suggestionFieldChange.mock.calls[0][0]).toEqual('AAA');
    expect(comp).toMatchSnapshot();
    expect(chunkUpdate.mock.calls.length).toEqual(0);
    comp.find('#add-stock').simulate('click');
    expect(chunkUpdate.mock.calls.length).toEqual(1);
    expect(comp).toMatchSnapshot();
  });

  it('add stock fail', () => {
    const suggestionFieldChange = jest.fn();
    const chunkUpdate = jest.fn();
    const comp = shallow(<InvestComposition
      chunks={[]}
      total={0.0}
      chunkUpdate={chunkUpdate}
      suggestionFieldChange={suggestionFieldChange}
      suggestions={[]}
      saveFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(suggestionFieldChange.mock.calls.length).toEqual(0);
    comp.find('#select-stock').simulate('change', { target: { value: 'AAA' } });
    comp.find('#add-stock').simulate('click');
    expect(chunkUpdate.mock.calls.length).toEqual(0);
  });

  it('delete stock', () => {
    const chunkUpdate = jest.fn();
    const comp = shallow((
      <InvestComposition
        chunks={[{
          id: 2, quantity: 1, value: 3, name: 'Name2',
        }, {
          id: 1, quantity: 1, value: 3, name: 'Name',
        }]}
        total={0.0}
        chunkUpdate={chunkUpdate}
        suggestionFieldChange={jest.fn()}
        suggestions={[]}
        saveFunc={jest.fn()}
        cancelFunc={jest.fn()}
      />
    ));
    expect(chunkUpdate.mock.calls.length).toEqual(0);
    comp.find('#delete-chunk-1').simulate('click');
    expect(chunkUpdate.mock.calls.length).toEqual(1);
    expect(chunkUpdate.mock.calls[0][0]).toEqual([{
      id: 2, quantity: 1, value: 3, name: 'Name2',
    }]);
  });

  it('change composition', () => {
    const chunkUpdate = jest.fn();
    const comp = shallow((
      <InvestComposition
        chunks={[{
          id: 1, quantity: 1, value: 3, name: 'Name',
        }, {
          id: 2, quantity: 2, value: 2, name: 'Name2',
        }]}
        total={100.0}
        chunkUpdate={chunkUpdate}
        suggestionFieldChange={jest.fn()}
        suggestions={[]}
        saveFunc={jest.fn()}
        cancelFunc={jest.fn()}
      />
    ));
    expect(chunkUpdate.mock.calls.length).toEqual(0);
    comp.find('#range').props().onChange([0, 6, 18]);
    expect(chunkUpdate.mock.calls.length).toEqual(1);
    expect(chunkUpdate.mock.calls[0][0]).toEqual([{
      id: 1, name: 'Name', quantity: 2, value: 3,
    }, {
      id: 2, name: 'Name2', quantity: 6, value: 2,
    }]);
    comp.find('#range').props().onChange([0, 6, 18]);
    expect(chunkUpdate.mock.calls.length).toEqual(1);
    comp.find('#range').props().onChange([0, 6, 16]);
  });
});
