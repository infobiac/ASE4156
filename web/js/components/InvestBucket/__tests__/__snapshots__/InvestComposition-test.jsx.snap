// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestComposition add stock 1`] = `
<withStyles(Dialog)
  open={true}
>
  <withStyles(DialogTitle)>
    Edit Composition
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(Table)>
      <withStyles(TableHead)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Delete
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Name
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Quantity
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Value
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Total Value
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableHead)>
      <withStyles(TableBody)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            <withStyles(IconButton)
              id="add-stock"
              onClick={[Function]}
            >
              <pure(Add) />
            </withStyles(IconButton)>
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            <Autocomplete
              autoHighlight={true}
              getItemValue={[Function]}
              id="select-stock"
              inputProps={Object {}}
              items={
                Array [
                  Object {
                    "id": 1,
                    "name": "AAA",
                    "value": 3,
                  },
                ]
              }
              menuStyle={
                Object {
                  "background": "rgba(255, 255, 255, 0.9)",
                  "borderRadius": "3px",
                  "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
                  "fontSize": "90%",
                  "maxHeight": "50%",
                  "overflow": "auto",
                  "padding": "2px 0",
                  "position": "fixed",
                }
              }
              onChange={[Function]}
              onMenuVisibilityChange={[Function]}
              onSelect={[Function]}
              renderInput={[Function]}
              renderItem={[Function]}
              renderMenu={[Function]}
              selectOnBlur={false}
              value="AAA"
              wrapperProps={Object {}}
              wrapperStyle={
                Object {
                  "display": "inline-block",
                }
              }
            />
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          >
            Available
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            numeric={true}
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableBody)>
    </withStyles(Table)>
    <ComponentEnhancer(Range)
      activeDotStyle={Object {}}
      allowCross={true}
      className=""
      count={0}
      disabled={false}
      dotStyle={Object {}}
      dots={false}
      handle={[Function]}
      handleStyle={
        Array [
          Object {},
        ]
      }
      id="range"
      included={true}
      marks={Object {}}
      max={0}
      min={0}
      onAfterChange={[Function]}
      onBeforeChange={[Function]}
      onChange={[Function]}
      prefixCls="rc-slider"
      pushable={false}
      railStyle={Object {}}
      step={0.01}
      trackStyle={
        Array [
          Object {},
        ]
      }
      value={
        Array [
          0,
        ]
      }
      vertical={false}
    />
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`InvestComposition add stock 2`] = `
<withStyles(Dialog)
  open={true}
>
  <withStyles(DialogTitle)>
    Edit Composition
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(Table)>
      <withStyles(TableHead)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Delete
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Name
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Quantity
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Value
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Total Value
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableHead)>
      <withStyles(TableBody)>
        <withStyles(TableRow)
          key="1"
        >
          <withStyles(TableCell)
            padding="dense"
          >
            <withStyles(IconButton)
              id="delete-chunk-1"
              onClick={[Function]}
            >
              <pure(DeleteForever) />
            </withStyles(IconButton)>
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            AAA
          </withStyles(TableCell)>
          <withStyles(TableCell)
            numeric={true}
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            numeric={true}
            padding="dense"
          >
            3
          </withStyles(TableCell)>
          <withStyles(TableCell)
            numeric={true}
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            <withStyles(IconButton)
              id="add-stock"
              onClick={[Function]}
            >
              <pure(Add) />
            </withStyles(IconButton)>
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            <Autocomplete
              autoHighlight={true}
              getItemValue={[Function]}
              id="select-stock"
              inputProps={Object {}}
              items={
                Array [
                  Object {
                    "id": 1,
                    "name": "AAA",
                    "value": 3,
                  },
                ]
              }
              menuStyle={
                Object {
                  "background": "rgba(255, 255, 255, 0.9)",
                  "borderRadius": "3px",
                  "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
                  "fontSize": "90%",
                  "maxHeight": "50%",
                  "overflow": "auto",
                  "padding": "2px 0",
                  "position": "fixed",
                }
              }
              onChange={[Function]}
              onMenuVisibilityChange={[Function]}
              onSelect={[Function]}
              renderInput={[Function]}
              renderItem={[Function]}
              renderMenu={[Function]}
              selectOnBlur={false}
              value=""
              wrapperProps={Object {}}
              wrapperStyle={
                Object {
                  "display": "inline-block",
                }
              }
            />
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          >
            Available
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            numeric={true}
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableBody)>
    </withStyles(Table)>
    <ComponentEnhancer(Range)
      activeDotStyle={Object {}}
      allowCross={true}
      className=""
      count={1}
      disabled={false}
      dotStyle={Object {}}
      dots={false}
      handle={[Function]}
      handleStyle={
        Array [
          Object {},
        ]
      }
      id="range"
      included={true}
      marks={Object {}}
      max={0}
      min={0}
      onAfterChange={[Function]}
      onBeforeChange={[Function]}
      onChange={[Function]}
      prefixCls="rc-slider"
      pushable={false}
      railStyle={Object {}}
      step={0.01}
      trackStyle={
        Array [
          Object {},
        ]
      }
      value={
        Array [
          0,
          0,
        ]
      }
      vertical={false}
    />
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`InvestComposition assert suggestion render 1`] = `
<TextField
  id="stockname"
  inputRef={null}
  required={false}
  select={false}
/>
`;

exports[`InvestComposition assert text render 1`] = `
<withStyles(Paper)
  style={
    Object {
      "background": "lightgray",
    }
  }
>
  testname
</withStyles(Paper)>
`;

exports[`InvestComposition assert text render 2`] = `
<withStyles(Paper)
  style={
    Object {
      "background": "white",
    }
  }
>
  testname
</withStyles(Paper)>
`;

exports[`InvestComposition render 1`] = `
<withStyles(Dialog)
  open={true}
>
  <withStyles(DialogTitle)>
    Edit Composition
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(Table)>
      <withStyles(TableHead)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Delete
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Name
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Quantity
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Value
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Total Value
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableHead)>
      <withStyles(TableBody)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            <withStyles(IconButton)
              id="add-stock"
              onClick={[Function]}
            >
              <pure(Add) />
            </withStyles(IconButton)>
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            <Autocomplete
              autoHighlight={true}
              getItemValue={[Function]}
              id="select-stock"
              inputProps={Object {}}
              items={Array []}
              menuStyle={
                Object {
                  "background": "rgba(255, 255, 255, 0.9)",
                  "borderRadius": "3px",
                  "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
                  "fontSize": "90%",
                  "maxHeight": "50%",
                  "overflow": "auto",
                  "padding": "2px 0",
                  "position": "fixed",
                }
              }
              onChange={[Function]}
              onMenuVisibilityChange={[Function]}
              onSelect={[Function]}
              renderInput={[Function]}
              renderItem={[Function]}
              renderMenu={[Function]}
              selectOnBlur={false}
              value=""
              wrapperProps={Object {}}
              wrapperStyle={
                Object {
                  "display": "inline-block",
                }
              }
            />
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          >
            Available
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            numeric={true}
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableBody)>
    </withStyles(Table)>
    <ComponentEnhancer(Range)
      activeDotStyle={Object {}}
      allowCross={true}
      className=""
      count={0}
      disabled={false}
      dotStyle={Object {}}
      dots={false}
      handle={[Function]}
      handleStyle={
        Array [
          Object {},
        ]
      }
      id="range"
      included={true}
      marks={Object {}}
      max={0}
      min={0}
      onAfterChange={[Function]}
      onBeforeChange={[Function]}
      onChange={[Function]}
      prefixCls="rc-slider"
      pushable={false}
      railStyle={Object {}}
      step={0.01}
      trackStyle={
        Array [
          Object {},
        ]
      }
      value={
        Array [
          0,
        ]
      }
      vertical={false}
    />
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;
